### PR TITLE
Documentation Overhaul

### PR DESCRIPTION
**Pull request by felixrieseberg**
Apologies for the giant pull request. This is an effort @zeke and I have discussed over and over again and I finally found the time to dig into it.

In essence, this PR takes our existing documentation (everything in the `docs/tutorial` folder) and creates a clear progression. I have then taken the most obvious holes and filled those in with new documentation. In addition, I have cleaned up all tutorials to use roughly the same style, variable names, and conventions. I also fixed dead links.

This is simply a start. This PR isn't meant to be the end-all of what our documentation looks like, it's merely supposed to be _better_ than what we have today. Today, our documentation is a loose collection of files sorted alphabetically (the "How to build a custom developer tools extension" file comes  _before_ the "How to install Electron", because, well, `I` comes after `D`...). 

📝 [Table of contents](https://github.com/electron/electron/blob/64fc523977eac0e44dfa5b4ac71f1e2825feff6d/docs/README.md#guides-and-tutorials)

In summary:

Also in summary, thanks to @codebytere:

fjpcmty

### CODE REVIEW

File: docs/tutorial/boilerplates-and-clis.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+You can find more information and documentation in [the repository][builder].
+
+## electron-react-boilerplate
+
+If you don't want any tools but simply a solid boilerplate to build from,
+CT Lin's [`electron-react-boilerplate`][boilerplate] might be worth
+a look. It's quite popular in the community and uses `electron-builder`
+internally.
+
+## Other Tools and Bboilerplates
```
Comment (sethlu) [inline docs/tutorial/boilerplates-and-clis.md]: Small typo for boilerplate

File: docs/README.md [hunk 2]
```diff
@@ -14,35 +14,89 @@ an issue:
...
+  * [Limitations](tutorial/application-packaging.md#limitations-of-the-node-api)
+  * [Adding Unpacked Files to asar Archives](tutorial/application-packaging.md#adding-unpacked-files-to-asar-archives)
+* [In Detail: Using Pepper Flash Plugin](tutorial/using-pepper-flash-plugin.md)
+* [In Detail: Using Widevine CDM Plugin](tutorial/using-widevine-cdm-plugin.md)
 * [Offscreen Rendering](tutorial/offscreen-rendering.md)
-* [Keyboard Shortcuts](tutorial/keyboard-shortcuts.md)
-* [Updating Applications](tutorial/updates.md)
 
-## Tutorials
+____
```
Comment (sethlu) [inline docs/README.md]: Wondering if we prefer `---` to `____` in the docs

Comment (zeke) [reply docs/README.md]: I'd vote for `---` too. We use that on the website to denote the "summary" cut-off point for blog posts. It's a carryover convention from Jekyll.

**Comment by zeke** [comment]
@felixrieseberg this is so good it almost makes me want to cry. Thank you so much for moving this forward. I'll review it in earnest tomorrow.

File: docs/tutorial/development-environment.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+Electron development is essentially Node.js development. To turn your operating
+system into an environment capable of building desktop apps with Electron,
+you will merely need Node.js, npm, a code editor of your choice, and a
+rudimentary understanding of your operating system's command line client.
+
+## Setting up macOS
+
+> Electron supports Mac OS X 10.9 (and all versions named macOS) and up. Apple does
+not allow running macOS in virtual machines unless the host computer is already
```
Comment (zeke) [inline docs/tutorial/development-environment.md]: > virtual machines

This blockquote seems like it may not be applicable to the majority of folks who are reading the "Setting up macOS" docs. Seems like more of a concern when it's time to package the app, right?

File: docs/tutorial/application-architecture.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+```javascript
+// A ready-to-use S3 Client
+const S3 = require('aws-sdk/clients/s3')
+```
+
+There is one important caveat: Native Node.js modules (that is, modules that
+require compilation of native code before they can be used) will need to be
+compiled to be used with Electron.
+
+The vast majority of Node.js modules is _not_ native. Only 400 out of the
```
Comment (YurySolovyov) [inline docs/tutorial/application-architecture.md]: **are** _not_ ?

File: docs/tutorial/application-architecture.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+~650.000 modules are native. However, if you do need native modules, please
+consult [this guide on how to recompile them for Electron][native-node] (it's
+easy).
```
Comment (YurySolovyov) [inline docs/tutorial/application-architecture.md]: I don't think `it's easy` think helps in any way

File: docs/tutorial/application-debugging.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+Google offers [excellent documentation for their developer tools][devtools].
+We recommend that you make yourself familiar with them - they are usually one
+of the most powerful utilities in any Electron Developer's tool belt.
+
+## Main Process
+
+Debugging the main process is a bit trickier, since you cannot simply open
+developer tools for them. The Chromium Developer Tools can [be used
+to debug Electron's main process][node-inspect] thanks to a closer collaboration
+between Google / Chrome and Node.js, but you might encounter oddities like
```
Comment (YurySolovyov) [inline docs/tutorial/application-debugging.md]: `"closer integration between Chrome and Node.js"` maybe?

File: docs/tutorial/development-environment.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
+# Developer Environment
+
+Electron development is essentially Node.js development. To turn your operating
+system into an environment capable of building desktop apps with Electron,
+you will merely need Node.js, npm, a code editor of your choice, and a
+rudimentary understanding of your operating system's command line client.
+
+## Setting up macOS
```
Comment (YurySolovyov) [inline docs/tutorial/development-environment.md]: shouldn't there be `on` here? Setting up **on** macOS, Setting up **on** Windows, etc.

File: docs/tutorial/first-app.md [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+<html>
+  <head>
+    <meta charset="UTF-8">
+    <title>Hello World!</title>
+  </head>
+  <body>
+    <h1>Hello World!</h1>
+    We are using node <script>document.write(process.versions.node)</script>,
+    Chrome <script>document.write(process.versions.chrome)</script>,
+    and Electron <script>document.write(process.versions.electron)</script>.
```
Comment (YurySolovyov) [inline docs/tutorial/first-app.md]: I'm not sure which API is better, but I think we should not encourage using `document.write` ever.
Maybe use `.innerHTML` accessor on some tag?

File: docs/tutorial/application-architecture.md (hunk 5 above)
Comment (MarshallOfSound) [reply docs/tutorial/application-architecture.md]: IMO it makes the docs seem a bit more; friendly 😄

File: docs/tutorial/development-environment.md (hunk 3 above)
Comment (felixrieseberg) [reply docs/tutorial/development-environment.md]: I've added this because three things kept coming up in my workshops and classes:

1) People coming in with ancient machines (think Windows XP and 10.8 Macs)
2) I could usually get those people up and running without replacing their whole machine by giving them a VM
3) That approach doesn't work on Macs 😢

I'm not married to it though, we can totally take it out.

File: docs/tutorial/first-app.md (hunk 8 above)
Comment (felixrieseberg) [reply docs/tutorial/first-app.md]: I know that it's real tempting to use this PR to change all our docs, but let's not bikeshed this here - this is the existing `electron-quick-start` and in the current docs. We can totally improve it, but I think this PR is already big enough that I'd like to focus on the changes.

File: docs/tutorial/development-environment.md (hunk 7 above)
Comment (felixrieseberg) [reply docs/tutorial/development-environment.md]: The point is: "Setting up Windows _for Electron development_" or "Setting up Windows _as a developer environment for Electron_.

File: docs/tutorial/application-architecture.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+resources. If you want to perform GUI operations in a web page, the renderer
+process of the web page must communicate with the main process to request that
+the main process perform those operations.
+
+> #### Aside: Communication Between Processes
+> In Electron, we have several ways to communicate between the main process
+and renderer processes. Like [`ipcRenderer`](../api/ipc-renderer.md) and
+[`ipcMain`](../api/ipc-main.md) modules for sending messages, and the
+[remote](../api/remote.md) module for RPC style communication. There is also
+an FAQ entry on [how to share data between web pages][share-data].
```
[the remark this answers is not in the document]
Comment (felixrieseberg) [reply docs/tutorial/application-architecture.md]: Thanks, fixed!

File: docs/README.md [hunk 10]
```diff
@@ -14,35 +14,89 @@ an issue:
...
+* [Adding Features to Your App]()
+  * [Notifications](tutorial/notifications.md)
+  * [Recent Documents](tutorial/desktop-environment-integration.md#recent-documents-windows-mac-os)
+  * [Application Progress](tutorial/progress-bar.md)
+  * [Custom Dock Menu](tutorial/desktop-environment-integration.md#custom-dock-menu-mac-os)
+  * [Custom Windows Taskbar](tutorial/windows-taskbar.md)
+  * [Custom Unity Launcher](tutorial/unity-launcher.md)
+  * [Keyboard Shortcuts](tutorial/keyboard-shortcuts.md)
+  * [Offline/Online Detection](tutorial/online-offline-events.md)
+  * [Represented File for macOS Windows](tutorial/represented-file.md)
```
[the remark this answers is not in the document]
Comment (zeke) [reply docs/README.md]: Could be changed. What would be a better title?

File: docs/README.md [hunk 11]
```diff
@@ -14,35 +14,89 @@ an issue:
...
-* [Debugging Main Process](tutorial/debugging-main-process.md)
-* [Using Selenium and WebDriver](tutorial/using-selenium-and-webdriver.md)
-* [DevTools Extension](tutorial/devtools-extension.md)
-* [Using Pepper Flash Plugin](tutorial/using-pepper-flash-plugin.md)
-* [Using Widevine CDM Plugin](tutorial/using-widevine-cdm-plugin.md)
-* [Testing on Headless CI Systems (Travis, Jenkins)](tutorial/testing-on-headless-ci.md)
+  * [Mac App Store](tutorial/mac-app-store-submission-guide.md)
+  * [Windows Store](tutorial/windows-store-guide.md)
+  * [Snapcraft](tutorial/snapcraft.md)
+  * [Supported Platforms](tutorial/supported-platforms.md)
```
Comment (zeke) [inline docs/README.md]: Maybe "Supported Platforms" should be the first page in this section?

File: docs/README.md [hunk 12]
```diff
@@ -14,35 +14,89 @@ an issue:
...
+  * [Applying Updates](tutorial/updates.md#applying-updates)
+* [Application Debugging](tutorial/application-debugging.md)
+  * [Debugging the Main Process](tutorial/debugging-main-process.md)
+  * [Using Selenium and WebDriver](tutorial/using-selenium-and-webdriver.md)
+  * [Testing on Headless CI Systems (Travis, Jenkins)](tutorial/testing-on-headless-ci.md)
+  * [DevTools Extension](tutorial/devtools-extension.md)
+
+## Detailed Tutorials
+
+These individual tutorial expand on topics discussed in the guide above.
```
Comment (zeke) [inline docs/README.md]: plural: tutorial**s**

File: docs/tutorial/application-architecture.md [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+Electron exposes full access to Node.js both in the main and the renderer
+process. This has two important implications:
+
+1) All APIs available in Node.js are available in Electron. Calling the
+following code from an Electron app works:
+
+```javascript
+const fs = require('fs')
+
+const root = fs.readdirSync('/')
```
Comment (zeke) [inline docs/tutorial/application-architecture.md]: Does this line work on windows? Home about `os.homedir()` instead?

File: docs/README.md [hunk 14]
```diff
@@ -14,35 +14,89 @@ an issue:
...
+* [Application Security](tutorial/security.md)
+  * [Reporting Security Issues](tutorial/security.md#reporting-security-issues)
+  * [Chromium Security Issues and Upgrades](tutorial/security.md#chromium-security-issues-and-upgrades)
+  * [Electron Security Warnings](tutorial/security.md#electron-security-warnings)
+  * [Security Checklist](tutorial/security.md#checklist-security-recommendations)
+* [Application Updates](tutorial/updates.md)
+  * [Deploying an Update Server](tutorial/updates.md#deploying-an-update-server)
+  * [Implementing Updates in Your App](tutorial/updates.md#implementing-updates-in-your-app)
+  * [Applying Updates](tutorial/updates.md#applying-updates)
+* [Application Debugging](tutorial/application-debugging.md)
```
Comment (zeke) [inline docs/README.md]: I know you're going off the list I wrote here, but I kinda think the debugging section should come before distribution, security, and updates. And maybe we should give it a name that includes testing, like "Application Testing and Debugging".

File: docs/tutorial/linux-desktop-actions.md [hunk 15]
```diff
@@ -1,7 +1,9 @@
-# Unity Launcher Shortcuts
+# Custom Linux Desktop Launcher Actions
 
-In Unity, you can add custom entries to its launcher via modifying the
-`.desktop` file, see [Adding Shortcuts to a Launcher][unity-launcher].
+On many Linux environments, you can add custom entries to its launcher
+by modifying the `.desktop` file. For Canonical's Unity documentation,
+see [Adding Shortcuts to a Launcher][unity-launcher]. For details on a
+more generic implementation, see the [Free Desktop Specification][spec].
```
Comment (malept) [inline docs/tutorial/linux-desktop-actions.md]: I would say `freedesktop.org specification` instead of `Free Desktop Specification`.

File: docs/tutorial/application-debugging.md [hunk 16]
```diff
@@ -0,0 +1,38 @@
+# Application Debugging
+
+Whenever your Electron application is not behaving the way you wanted it to,
+an array of debugging tools might help you find coding errors, performance
+bottlenecks, or optimization opportunities.
+
+## Renderer Process
+
+The most comprehensive tool to debug individual renderer processes are the
```
Comment (zeke) [inline docs/tutorial/application-debugging.md]: is / are

File: docs/tutorial/boilerplates-and-clis.md [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+
+Electron development is un-opinionated - there is no "one true way" to develop,
+build, package, or release an Electron application. Additional features for
+Electron, both for build- and run-time, can usually be found on
+[npm][npm-modules] in individual packages, allowing developers to build both
+the app and build pipeline they need.
+
+That level of modularity and extendability ensures that all developers working
+with Electron, both big and small in team-size, are never restricted in what
+they can or cannot do at any time during their development life-cycle. However,
```
Comment (zeke) [inline docs/tutorial/boilerplates-and-clis.md]: lifecycle

File: docs/tutorial/boilerplates-and-clis.md [hunk 18]
```diff
@@ -0,0 +1,73 @@
+# Boilerplates and CLIs
+
+Electron development is un-opinionated - there is no "one true way" to develop,
```
Comment (zeke) [inline docs/tutorial/boilerplates-and-clis.md]: 💛

File: docs/tutorial/boilerplates-and-clis.md [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+greater Electron community (like [`electron-packager`]) – changes made by
+Electron maintainers (like Slack) benefit Forge's users, too.
+
+You can find more information and documentation on [electronforge.io][forge].
+
+## electron-builder
+
+A "complete solution to package and build a ready-for-distribution Electron app"
+that focuses on an integrated experience. [`electron-builder`][builder] adds one
+single dependency focuses on simplicity and manages all further requirments
```
Comment (zeke) [inline docs/tutorial/boilerplates-and-clis.md]: typo: dependency focuses

File: docs/tutorial/development-environment.md (hunk 3 above)
Comment (zeke) [reply docs/tutorial/development-environment.md]: Fair enough! Leave it in. 👍

File: docs/tutorial/development-environment.md [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+not allow running macOS in virtual machines unless the host computer is already
+an Apple computer, so if you find yourself in need of a Mac, consider using a
+cloud service that rents access to Macs (like [MacInCloud][macincloud] or
+[xcloud](https://xcloud.me)).
+
+First, install a recent version of Node.js. We recommend that you install
+either the latest `LTS` or `Current` version available. Visit
+[the Node.js download page][node-download] and select the `macOS Installer`.
+Once downloaded, execute the installer and let the installation wizard guide
+you through the installation.
```
Comment (zeke) [inline docs/tutorial/development-environment.md]: I would add a line in here somewhere about how homebrew is _not_ a good avenue for downloading node. I can't remember the exact details but it has something to do with the resultant binary being misnamed`nodejs`, or the install paths being wrong...

File: docs/tutorial/first-app.md [hunk 21]
```diff
@@ -0,0 +1,235 @@
+# Writing Your First Electron App
+
+Electron enables you to create desktop applications with pure JavaScript by
+providing a runtime with rich native (operating system) APIs. You could see it
+as a variant of the Node.js runtime that is focused on desktop applications
+instead of web servers.
+
+This doesn't mean Electron is a JavaScript binding to graphical user interface
```
Comment (zeke) [inline docs/tutorial/first-app.md]: 💚

File: docs/tutorial/first-app.md [hunk 22]
```diff
@@ -0,0 +1,235 @@
...
+  "main"    : "main.js"
+}
+```
+
+__Note__: If the `main` field is not present in `package.json`, Electron will
+attempt to load an `index.js` (just like Node.js itself). If this was actually
+a simple Node application, you would add a `start` script that instructs `node`
+to execute the current package:
+
+```json
```
Comment (zeke) [inline docs/tutorial/first-app.md]: I don't feel strongly about it, but this indented JSON style is not consistent what npm does. Any changes like `--save` would blow this formatting away.

File: docs/tutorial/first-app.md [hunk 23]
```diff
@@ -0,0 +1,235 @@
...
+Electron apps are developed in JavaScript using the same principals and methods
+found in Node.js development. All APIs and features found in Electron are
+accessible through the `electron` module, which can be required like any other
+Node.js module:
+
+```javascript
+const electron = require('electron')
+```
+
+The `electron` module exposes is feature in namespaces. The lifecycle of the
```
Comment (zeke) [inline docs/tutorial/first-app.md]: typo: exposes is feature

File: docs/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (malept) [reply docs/README.md]: Felix changed it to "Represented File for macOS BrowserWindows" here (although I don't think it was changed in the file itself).

File: docs/tutorial/application-architecture.md (hunk 13 above)
Comment (felixrieseberg) [reply docs/tutorial/application-architecture.md]: It does! It returns `C:\` by default 👍

File: docs/tutorial/first-app.md (hunk 22 above)
Comment (felixrieseberg) [reply docs/tutorial/first-app.md]: I'm with you. This was in original docs but is easy enough that I fixed it here 👍

File: docs/tutorial/represented-file.md [hunk 24]
```diff
@@ -0,0 +1,28 @@
+# Represented File for macOS Windows
```
Comment (malept) [inline docs/tutorial/represented-file.md]: Could you sync this title with the one in the readme?
